src/Razor/src/Microsoft.AspNetCore.Razor.VSCode.BlazorWasmCompanionExtension/src/LocalDebugProxyManager.ts

```diff
@@ -0,0 +1,65 @@
+/* --------------------------------------------------------------------------------------------
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License. See License.txt in the project root for license information.
+ * ------------------------------------------------------------------------------------------ */
+
+const fetch = require('node-fetch');
+import * as stream from 'stream';
+import * as extract from 'extract-zip';
+import * as net from 'net';
+import * as os from 'os';
+import * as fs from 'fs';
+import * as util from 'util';
+
+export class LocalDebugProxyManager {
+    private readonly localPath = `${os.homedir()}/.blazor-wasm-debug-proxy-local`;
+    private readonly nugetUrl = 'https://api.nuget.org/v3-flatcontainer';
+    private readonly packageName = 'Microsoft.AspNetCore.Components.WebAssembly.DevServer';
+    private readonly finished = util.promisify(stream.finished);
+
+    constructor() {
+        if (!fs.existsSync(this.localPath)) {
+            fs.mkdirSync(this.localPath);
+        }
+    }
+
+    public async getDebugProxyLocalNugetPath(version: string) {
```

Comment:
Yep, unfortunately, since this logic must live in a different extension and run in the UI, it is difficult to reuse the code that already exists in the C# extension to solve this problem.

I had considered implementing something similar to that infrastructure in this extension, but it seemed a little overkill to include it for downloading a single package from NuGet. With regard to the points mentioned:

- Size-wise the package is about 4MB. Since we cache it on the users disk, it's a one-time install.
- We install the package via the NuGet API so we are bound by its availability.
- Similar the O# extension, assets are saved in a hidden directory in the users homedrive.

Reply:
Hmm, when I wrote the comment I didn't realize you were intending it to be its own extension. Some questions in regards to that then:

- Why does this need to be a different extension? 
- Don't we still need the C# extension in order to get a lot of the C# coloring behavior in VSCode even in a Codespaces environment? 
- How would this extension be "installed" on a guest box? Aka how would the user know to install it?

Might be easier to sit down and talk it through together, I think there's a few pieces of the puzzle I'm missing

Reply:
I wrote a little writeup on the design [here](https://github.com/dotnet/aspnetcore-tooling/blob/safia/remote-dbg/src/Razor/src/Microsoft.AspNetCore.Razor.VSCode.BlazorWasmCompanionExtension/README.md). This might help answer some of your questions. If not, we can sit down and chat and I'll update the doc based on our conversation.

> Don't we still need the C# extension in order to get a lot of the C# coloring behavior in VSCode even in a Codespaces environment?

Yep, the C# extension doesn't go away in this case. We're introducing a second companion extension that gets installed in the users VS Code UI context.

> How would this extension be "installed" on a guest box? Aka how would the user know to install it (would it be bundled with something else)?

The extensions API allows you to set a `extensionDependencies` property that lists the IDs of extensions that should be installed when a primary extension is so we can list this as an extensionDependency of the C# extension (preferred).

Alternatively, we can use the model we used when we had a dependency on the JS debugger extension and write some logic that alerts the user when they don't have it installed and prompts them to install.

Note: this PR doesn't contain any of these acquisition related changes since I wanted to get the extension logic verified first.

Reply:
Lets sit down and spend a little time talking about the details, I have some concerns